release-23.2: copy: improve behavior of non-atomic COPY

yuzefovich opened the pull request:
Backport 3/3 commits from #116512.

/cc @cockroachdb/release

---

This commit audits the behavior of non-atomic COPY. As a reminder, in such a setup we create a separate txn for each batch of rows, and the responsibility of committing these txns lies with the COPY state machine. To further complicate the story, we have a retry loop around each batch of rows, where we give each batch up to 5 times to be processed in case it is hitting retryable errors.

One problem that this commit addresses is how we reset the buffered data. Previously, we would do the reset right after executing the INSERT pipeline. In some cases, that pipeline results in the txn being auto committed, so we're good then; however, if the INSERTs cannot auto commit, then it is possible that we will fail to commit the txn in the COPY state machine, but by that point we will have reset the buffered data, effectively silently losing it. On the second retry attempt for this batch we'd have zero rows to insert, so in absence of other problems, we could end up proceeding with the COPY without inserting some of the data. This commit addresses this scenario by resetting the buffered state only after we know that the INSERTs have been committed.

However, I believe previously it wouldn't be possible to hit this silent data loss scenario because of a second issue this commit addresses. In particular, when the COPY state machine fails to commit its txn, it attempts to roll it back, and previously, regardless of whether the rollback succeeded or not, we would always short-circuit execution of the `cleanup` function returned by `preparePlannerForCopy` so that we'd end up keeping the same txn, that is in finalized state and couldn't be used anymore. Thus, on the next retry attempt we'd always get a non-retrible error failing the COPY overall. This commit addresses this issue by not short-circuiting the execution of that deferred method which creates a fresh txn for the next retry attempt.

It also now makes it so that if a txn rollback fails, then that non-retriable error is returned to the client, failing the COPY. This seems like a sane thing to me because at that point we're kind of in an undefined state. However, I could also see a benefit of proceeding still since we're in non-atomic COPY scenario and perhaps the user wants COPY to succeed even if it results in somewhat corrupted inserted data.

Kudos to Nathan for identifying these two problems.

As an additional minor change it fixes the number of inserted rows that could previously be inflated in case non-atomic COPY had internal retries (although, again, due to the second issue described above, this was hidden).

Epic: None

Release note (bug fix): CockroachDB can now retry more retryable errors transparently when performing non-atomic COPY command.

Release justification: bug fix.

yuzefovich committed:
copy: short-circuit earlier

This commit makes the logic in `insertRowsInternal` short-circuit the
execution earlier in case we have no rows to insert. In particular, this
allows us to avoid possibly creating a new txn that won't be used at
all.

Release note: None

yuzefovich committed:
copy: improve behavior of non-atomic COPY

This commit audits the behavior of non-atomic COPY. As a reminder, in
such a setup we create a separate txn for each batch of rows, and
the responsibility of committing these txns lies with the COPY state
machine. To further complicate the story, we have a retry loop around
each batch of rows, where we give each batch up to 5 times to be
processed in case it is hitting retryable errors.

One problem that this commit addresses is how we reset the buffered
data. Previously, we would do the reset right after executing the INSERT
pipeline. In some cases, that pipeline results in the txn being auto
committed, so we're good then; however, if the INSERTs cannot auto
commit, then it is possible that we will fail to commit the txn in the
COPY state machine, but by that point we will have reset the buffered
data, effectively silently losing it. On the second retry attempt for
this batch we'd have zero rows to insert, so in absence of other
problems, we could end up proceeding with the COPY without inserting
some of the data. This commit addresses this scenario by resetting the
buffered state only after we know that the INSERTs have been committed.

However, I believe previously it wouldn't be possible to hit this silent
data loss scenario because of a second issue this commit addresses. In
particular, when the COPY state machine fails to commit its txn, it
attempts to roll it back, and previously, regardless of whether the
rollback succeeded or not, we would always short-circuit execution of
the `cleanup` function returned by `preparePlannerForCopy` so that we'd
end up keeping the same txn, that is in finalized state and couldn't be
used anymore. Thus, on the next retry attempt we'd always get
a non-retrible error failing the COPY overall. This commit addresses
this issue by not short-circuiting the execution of that deferred
method which creates a fresh txn for the next retry attempt.

It also now makes it so that if a txn rollback fails, then that
non-retriable error is returned to the client, failing the COPY. This
seems like a sane thing to me because at that point we're kind of in an
undefined state. However, I could also see a benefit of proceeding still
since we're in non-atomic COPY scenario and perhaps the user wants COPY
to succeed even if it results in somewhat corrupted inserted data.

Kudos to Nathan for identifying these two problems.

As an additional minor change it fixes the number of inserted rows that
could previously be inflated in case non-atomic COPY had internal
retries (although, again, due to the second issue described above, this
was hidden).

Release note (bug fix): CockroachDB can now retry more retryable errors
transparently when performing non-atomic COPY command.

yuzefovich committed:
copy: add a session variable for the number of retries per batch

Previously, for every batch of the non-atomic COPY would get retried up
to 5 times (meaning there would be at most 6 attempts). This commit
introduces a session variable that can control this number - this might
come handy in the future as an additional knob to make COPY succeed
(e.g. in the context of DMS).

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/117895)
<!-- Reviewable:end -->

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r1, 5 of 5 files at r2, 7 of 7 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/117895#-:-NoOZ7_w4mgfRtQyiDyQ:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mgartner and @rafiss)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
The only conflict was very minor.

> Are all of the changes protected via a flag or option, new syntax, an environment variable or default-disabled cluster setting?

Not exactly, but the changes here only affect non-atomic copy, which is non-default behavior.

> What are the risks to backporting this change?

The main risk I can see is that we somehow messed up txn commit / rollback logic so that non-atomic COPY somehow gets stuck (either in an infinite loop due to retries or because it can never commits its batches).

> Can the risks of backporting be mitigated?

The changes to existing txn logic were as minor as possible. The number of retries is now configurable too.

> What are the risks to not backporting?

The customers using COPY (including DMS) will keep on running into difficulties with migrations.

> Does this change really need to be backported? Or can it reasonably wait until the next major release to be addressed?

Backporting this change should noticeably reduce L2 escalations, so the backport is of high benefit.

mgartner reviewed (approved):
LGTM

yuzefovich committed:
copy: deflake TestCopyFromRetries

`TestCopyFromRetries` can take very long time when
`sql.CopyBatchRowSize` is set to a large value. This constant is 100 by
default but can be randomized up to 50k. In one test run this constant
was set to 48k and the test timed out in one of multi-tenant configs.
There is no much benefit from running this test with such a large COPY
batch size, so this commit caps it at 1k.

Release note: None

yuzefovich commented:
Included test fixup from #118135.
